nemo/collections/nlp/models/information_retrieval/bert_dpr_model.py

```diff
@@ -0,0 +1,277 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import Dict, Optional
+
+import numpy as np
+import torch
+from omegaconf import DictConfig
+from pytorch_lightning import Trainer
+
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.nlp.data import BertInformationRetrievalDatasetEval, BertInformationRetrievalDatasetTrain
+from nemo.collections.nlp.modules.common.lm_utils import get_lm_model
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_tokenizer
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.modelPT import ModelPT
+from nemo.core.neural_types import ChannelType, LogitsType, MaskType, NeuralType
+
+__all__ = ['BertDPRModel']
+
+
+class BertDPRModel(ModelPT):
+    """
+    Information retrieval model which encodes query and passage separately
+    with two different BERT encoders and computes their similarity score
+    as a dot-product between corresponding [CLS] token representations.
+    """
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "q_input_ids": NeuralType(('B', 'T'), ChannelType()),
+            "q_attention_mask": NeuralType(('B', 'T'), MaskType()),
+            "q_token_type_ids": NeuralType(('B', 'T'), ChannelType()),
+            "p_input_ids": NeuralType(('B', 'T'), ChannelType()),
+            "p_attention_mask": NeuralType(('B', 'T'), MaskType()),
+            "p_token_type_ids": NeuralType(('B', 'T'), ChannelType()),
+        }
+
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {"logits": NeuralType(('B', 'D'), LogitsType())}
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+
+        self.dataset_cfg = cfg.dataset
+
+        self.tokenizer = get_tokenizer(tokenizer_name=cfg.language_model.pretrained_model_name)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.q_encoder = get_lm_model(pretrained_model_name=cfg.language_model.pretrained_model_name)
```

Comment:
please use it like https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/question_answering/qa_model.py#L53-L58
since we want to make it general enough to support other lm's and tokenizers